Add test action for aws-robomaker-github-actions

wangvnn opened the pull request:
- Add integration test job for robomaker-sample-app-ci
- Steps: 
   + Setup ros environment (linux distro, ros distro, gazebo)
   + Checkout hello world SA
   + Check out target action
   + Setup nodeJS environment (packages, build)
   + Run target action: robomaker-sample-app-ci
   + Verify expect output files exists: bundle files, source files, build artifacts, dependencies
- Also improve robomaker-sample-app-ci script to work with long working-dir (multiple directories)

wangvnn committed:
Kenvink/test actions (#8)

Add test action for aws-robomaker-github-actions
- Add integration test job for robomaker-sample-app-ci
- Steps: 
   + Setup ros environment (linux distro, ros distro, gazebo)
   + Check out target action
   + Setup nodeJS environment (packages, build)
   + Run target action: robomaker-sample-app-ci
   + Verify expect output files exists: bundle files, source files, build artifacts, dependencies
- Also improve robomaker-sample-app-ci script to work with long working-dir (multiple directories)

wangvnn committed:
Change to run on master and release branches

ojasjoshi commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,106 @@
+      
+    # Trigger to run robomaker-sample-app-ci action script
+    - name: Build and bundle robot_ws      
+      run: node ./robomaker-sample-app-ci/dist/index.js
```
Do we need to add tests for other actions as well?

wangvnn replied:
For 1st iteration, I think we should run tests for robomaker-sample-app-ci first since it is the most complex action in this repos. And yes we can add more tests for other actions as well as needed. I prefer to do that in another PR.

ojasjoshi commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,106 @@
+      env:
+        INPUT_ROS-DISTRO: ${{ matrix.distro }}
+        INPUT_GAZEBO-VERSION: ${{ matrix.gazebo }}
+        INPUT_WORKSPACE-DIR: ./.github/workflows/data/robot_ws
```
Should we be fetching the latest source files for build and bundle from the SA github repos? I think it's a bit tricky as changes in SA repos should not break the CI tests for these actions but as we last discussed, any breaking changes would <em>never be pushed</em> to the head of SA repos. With current setup, there is a chance that these tests might get redundant if the SA repo changes.

wangvnn replied:
I tested with SA..but here I am trying to minimize the number of dependencies. Integration test itself already has enough complexity (run docker image, install ros...) the more dependencies we have the higher chance it will break. So I decided to include very simple example app embedded in the test data instead. The goal of the tests to verify that the action itself working as expected, not to tight it to any SA.

ojasjoshi replied:
Yes I agree with decoupling these tests and actions from SA's.. robomaker-sample-app-ci action in itself is however not decoupled entirely (for eg: [getSampleVersion()](https://github.com/aws-robotics/aws-robomaker-github-actions/blob/03720c02e1af2052ba7a59140c0b00b17d302ac2/robomaker-sample-app-ci/src/aws-robomaker-sample-application-ci.ts#L52))..so just want to raise the question -- in case a breaking change is made in a coupled function within the actions here, do we want the CI to break here or in the Sample App? Or do we not check such SA coupled changes in the CI tests?

wangvnn replied:
To me, SA is just general term we are using for multiple sample app(s) that have this structure: robot_ws, simulation_ws... (aws-robomaker-sample-application-helloworld, aws-robomaker-sample-application-cloudwatch). The embedded SA follows that structure, so getSampleVersion is not a problem. If there is breaking change in this repos, the CI of this repos should catch it. This is also good question because it raises another issue that was addressed in another discussion (stability and security). I think if we really want to address the stability, we should use versioning tag (or commit SHA) as  recommended by Github (making change in action, test it, tag it as major/minor version... and make change to SA as needed if SA's CI wants to use newer versions)

ojasjoshi replied:
As far as my knowledge, we do have versioning tags for actions and the corresponding version tags are being used in the GH action workflows in Cloudwatch and Helloworld SAs.Is there a benefit to using commit SHA over version tags?

mm318 replied:
With my [comment above](https://github.com/aws-robotics/aws-robomaker-github-actions/pull/9/files#r488856420), I would prefer us using our sample apps for this.

@wangvnn if you're concerned with sample apps constantly changing such that there's too many moving parts to this test, you can fetch a specific tag or commit of the sample apps instead.

wangvnn replied:
The issue is not about which version of SA...but the complexity of SA. SA can do more things than we need (e.g pull in turtle bot source from github) if that fails we fails our CI here and that is unnecessary.

mm318 replied:
Pulling TurtleBot sources (executing `vcs import`) is the responsibility of this GitHub action. If you remove that complexity, I think you won't be testing this GitHub action sufficiently.

wangvnn replied:
Yep, I think I missed that vcs import. I was even thinking about a simpler .rosinstall to pull in something simpler than turtlebot. 
It looks like team prefer pulling the SA. I will make the change. Which one do you prefer for this test to pull in (hello world or cloud watch or both) ?

mm318 replied:
In terms of code coverage, I think either sample app will exercise the same amount of code in this GitHub action. Let's go with just the Hello World sample app, since it should build and bundle slightly quicker.

wangvnn replied:
new update: using hello world app for testing now.

ojasjoshi commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,106 @@
+    # Expectation: bundle file, build files, and dependencies all exist
+    - name: Check robot_ws file existence
+      id: check_robot_ws_files
+      uses: andstor/file-existence-action@v1
```
Would using our own fork for this third party repo be a better idea? If not, any breaking changes to the third party action would be propagated here.

wangvnn replied:
That is good point. Github recommended this: "Using the commit SHA of a released action version is the safest for stability and security"
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsuses
I will make a change to include the commit SHA to address stability and security concern.

ojasjoshi replied:
Yep that looks good!

mm318 replied:
Hmm, I just made a comment about this [here](https://github.com/aws-robotics/aws-robomaker-github-actions/pull/9#discussion_r488879471).

I think using tags is a better convention. I understand that maintainers can force push to move or remove tags, but in the same vein they can overwrite commits, too, and the commit hash that you're pointing will no longer exist. I think tags and commit hash have the same level of stability.

We should trust the maintainer `andstor` and take a presumption of innocence (consider the maintainer to be innocent until proven guilty).

wangvnn replied:
This is what is recommended by Github. 
See it here: https://github.community/t/are-github-actions-safe-to-use/17895

mm318 replied:
Okay, I see. Their rationale is that we should rather use a hash commit that would no longer exist after a malicious force push than a branch or tag that would be pointing to malicious code after a malicious force push.

wangvnn committed:
Update test.yml

Use andstor/file-existence-action@87d74d4 instead of andstor/file-existence-action@v1 to address stability and security concern as it is recommended by Github:
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsuses

mm318 commented on .github/workflows/data/LICENSE:
```diff
@@ -0,0 +1,14 @@
+Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
I think the contents of this file should be the full license, not just the license header.

wangvnn replied:
Will remove the embedded app as discussed in another thread.

mm318 commented on .github/workflows/data/robot_ws/src/m2wr_description/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+cmake_minimum_required(VERSION 2.8.3)
```
I don't feel comfortable putting a copy other people's work our repos, especially when it's not explicitly licensed (https://github.com/mattborghi/m2wr_description).

I would prefer we stick to using our own work, like our sample apps.

Even when using our sample apps, prefer fetching than storing a copy of it in this repo.

mm318 replied:
Also, when fetching our sample apps, fetch it to the [default destination directory](https://github.com/aws-robotics/aws-robomaker-sample-application-helloworld/blob/1d8342a8dd4bc74f8a3f17b3192aec4bf1357cb9/.github/workflows/ros1.yml#L39), so that this repo's test workflow YAML will resemble more closely to that of a user's.

wangvnn replied:
When I tried this, action checkout will fails if we checkout another git repos. The clean solution for multiple repos is to checkout them into different folders.

mm318 replied:
You can specify the path where you fetch/checkout to: https://github.com/actions/checkout#checkout-multiple-repos-side-by-side

I ask that you fetch/checkout the sample apps to the default location, which means the GitHub action will need to be checked out to a non-default location. (The repo that's being checked out to the default location may need to be the first one that's checked out, ie. the sample app will need to be checked out before the GitHub action.)

mm318 commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,106 @@
+    - name: Build and bundle robot_ws      
+      run: node ./robomaker-sample-app-ci/dist/index.js
+      env:
+        INPUT_ROS-DISTRO: ${{ matrix.distro }}
```
What are these environment variables used for?

EDIT: I'm guessing it's `uses`/`with` versus `run`/`env`.

mm318 replied:
It looks like what `action-ros-ci` does is build (like you did with `npm run build`) and then do [`uses: ./`](https://github.com/ros-tooling/action-ros-ci/blob/master/.github/workflows/test.yml#L55). With `uses`, you can do the usual `with` statements, which is a lot cleaner than setting cryptic environment variables with `env`.

wangvnn replied:
I will test this again, in my first attempt, it did not build the latest from source so I change to run node directly instead.

wangvnn replied:
Update: remove custom node run, use "use: ./" as suggested now.

mm318 commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,106 @@
+    # Expectation: bundle file, build files, and dependencies all exist
+    - name: Check robot_ws file existence
+      id: check_robot_ws_files
+      uses: andstor/file-existence-action@87d74d4
```
Can we use a version tag (eg. `v1.0.1`) instead of a commit hash?

mm318 replied:
I'll accept the rational [here](https://github.com/aws-robotics/aws-robomaker-github-actions/pull/9#discussion_r488915550).

mm318 commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,106 @@
+      matrix:
+        distro: ['kinetic', 'melodic']
+        gazebo: [7, 9]
+        include:
```
Remember to test ROS2 Dashing as well (add a new workflow YAML file, like `test_ros2.yml`).

wangvnn replied:
I thought we discussed before we want to run this for ros 1 only.

mm318 replied:
Okay, let's add ROS2 Dashing as a test in a separate. We should do it at some point.

mm318 commented on robomaker-sample-app-ci/src/aws-robomaker-sample-application-ci.ts:
```diff
@@ -157,10 +157,15 @@ async function prepare_sources() {
-
+    
+    const execOptions: ExecOptions = {
+      cwd: path.join(WORKSPACE_DIRECTORY, ".."),
```
If [this](https://github.com/aws-robotics/aws-robomaker-github-actions/pull/9/files#r488882730) is followed, are these changes still needed?

I also don't think we can always assume that the workspace is `..` relative to the GitHub action's working directory. I'm actually concerned that this change will break our current sample apps' CI.

wangvnn replied:
There are a lot of assumptions in the action itself. Indeed, it did not support multiple folders in working-dir since it use working-dir for the bundle filename. Here, I make an assumption that we want to store the bundle file one level before the working-dir. If this assumption is not valid then we need to add other params to the input, something like: bundle-folder and source-folder... There is an assumption that the default root destination is at the github-action checkout folder which does not work well when checking out multiple repos.

wangvnn replied:
Update: refactor getExecOptions and add getWorkingDirExecOptions and getWorkingDirParentExecOptions to make sure we are clear with which folder we want to run the command.

wangvnn committed:
Update test.yml

Use action "use" instead of calling node to trigger action code.

wangvnn committed:
Update test.yml

wangvnn committed:
Update test.yml

Check out aws-robomaker-sample-application-helloworld as testing SA.

wangvnn committed:
Update test.yml

Checkout sample app hello world's branch ros1

wangvnn committed:
Update test.yml

Use checkout@v2 instead of v1

wangvnn committed:
Update test.yml

Move setup permission up since checkout to folder requires permission to create folder.

wangvnn committed:
Update test.yml

Fix a typo: generate-sources

wangvnn committed:
Update test.yml

Fix wrong working-dir

wangvnn committed:
Refactor getExecOptions to support relative path

- For bundling and zipping source file, we store those files outside of working folder
- Add getWorkingDirExecOptions and getWorkingDirParentExecOptions to make it clear which one we are using

wangvnn committed:
Update test.yml

Check out SA to default workspace folder as suggested

wangvnn committed:
Update test.yml

Just checkout the latest SA for now

mm318 commented on robomaker-sample-app-ci/src/aws-robomaker-sample-application-ci.ts:
```diff
@@ -179,8 +184,9 @@ async function build() {
 
 async function bundle() {
   try {
+    const bundleFilename = WORKSPACE_DIRECTORY.split(path.sep).pop()    
```
Is there a `basename()` function instead of doing a `split()` and `pop()`?

wangvnn replied:
Good point, testing basename now.

wangvnn replied:
Update: using basename now.

mm318 commented on robomaker-sample-app-ci/dist/index.js:
```diff
@@ -816,9 +816,13 @@ function prepare_sources() {
                 "README*",
                 "roboMakerSettings.json"
             ];
+            const execOptions = {
```
Looks like your `index.js` is out of sync with your `aws-robomaker-sample-application-ci.ts`.

wangvnn replied:
> Looks like your `index.js` is out of sync with your `aws-robomaker-sample-application-ci.ts`.

Keep it out of sync to test the npm run build....will check it in as final step.

wangvnn replied:
Update: index.js is in sync now.

wangvnn committed:
Clean up unused test data  and rebuild
- Remove embedded test app
- Rebuild .ts code to index.js

mm318 reviewed (approved):
Looks great!
